src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild.Rpc;
+
+/// <summary>
+/// Implements the server side of the RPC channel used to communicate with the build host.
+/// </summary>
+/// <remarks>
+/// The RPC system implemented here is pretty close to something like JSON-RPC; however since we need the Build Host to be usable in Source Build
+/// scenarios, we are limited to using only what is either in .NET or can be easily made buildable in Source Build. Thus existing solutions like StreamJsonRpc 
+/// are out. If at some point there is a standard RPC mechanism exposed in .NET or Source Build, we should delete this and use that instead.
+/// </remarks>
+internal sealed class RpcServer
+{
+    private readonly TextWriter _sendingStream;
+    private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+    private readonly TextReader _receivingStream;
+
+    private readonly ConcurrentDictionary<int, object> _rpcTargets = new();
+    private volatile int _nextRpcTargetIndex = -1; // We'll start at -1 so the first value becomes zero
+
+    private readonly CancellationTokenSource _shutdownTokenSource = new CancellationTokenSource();
+
+    public RpcServer(Stream sendingStream, Stream receivingStream)
+    {
+        _sendingStream = new StreamWriter(sendingStream, Encoding.UTF8);
+        _receivingStream = new StreamReader(receivingStream, Encoding.UTF8);
+    }
+
+    public int AddTarget(object rpcTarget)
+    {
+        // Loop until we successfully have a new index for this; practically we don't expect this to ever collide, since that'd mean we'd have
+        // billions of long lived projects, but...
+        while (true)
+        {
+            var nextIndex = Interlocked.Increment(ref _nextRpcTargetIndex);
+            if (_rpcTargets.TryAdd(nextIndex, rpcTarget))
```

Comment:
I'm probably just dumb, but I'm not understanding this loop.  Won't `nextIndex` always be unique?  Why do we need to loop here.

Reply:
We really should add removal and disposal for the project objects too, at which point a long-lived process could theoretically reuse them. This otherwise just copy/paste from the other target code that's doing the same thing.